File: search/TwoCrystalBalls.java
package search;

public class TwoCrystalBalls {

    public static int twoCrystalBalls(int[] breaks) {
        int jumps = (int) Math.floor(Math.sqrt(breaks.length));

        // jump and see if breaks
        int i = jumps;
        for (; i < breaks.length; i += jumps) {
            if (breaks[i] == 1) {
                break;
            }
        }

        // walk back one increment
        i = i + 1 - jumps;
        for (int j = 0; j < jumps && i < breaks.length; j++, i++) {
            if (breaks[i] == 1) {
                return i;
            }
        }

        // if not
        return -1;

    }

    public static void main(String[] args) {

        int[] breakingPoint = new int[] { 0, 0, 0, 0, 0, 0, 1, 1, 1, 1 };
        // print result
        System.out.println(twoCrystalBalls(breakingPoint));
    }

}

File: search/TwoCrystalBalls.js
const twoCrystalBalls = (breakingPoint) => {
    const jumps = Math.floor(Math.sqrt(breakingPoint.length));
    console.log("jump amount: " + jumps)

    // jump until the first ball breaks
    let i = jumps;
    for (; i < breakingPoint.length; i += jumps) {
        if (breakingPoint[i]) {
            console.log("Broken!");
            break;
        }
        console.log("Not broken");
    }

    // walk back one increment
    i = i - jumps + 1;
    // check one by one
    for (j = 0; j < jumps && i < breakingPoint.length; i++, j++) {
        if (breakingPoint[i]) {
            console.log("broken at: " + i)
            return i;
        }
        console.log("current position: " + i);
    }

    return -1;
}

const breakingPoint = [0, 0, 0, 0, 0, 0, 1, 1, 1, 1];
console.log(twoCrystalBalls(breakingPoint))